### DISCUSSION TITLE
Object-Oriented Programming이 무엇인가요?

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#42] [프론트 기술 질문 공간]
아래의 OOP 네가지 원칙에 대해 알려주세요.

**Comment by michoball** [comment]
Inheritance : 부모 객체가 가지고 있는 속성이 자식 객체가 사용할 수 있는 것 
Encapsulation: 특정 일에 관련된 기능과 값을 하나로 묶어두는 것 ( class 안에 넣어 두는 것)
Abstraction: 어떤 객체의 기능을 사용할 때 사용하는 쪽에서는 내부에 어떤 일이 일어나는지 알 필요없이 이름만으로 기능을 알 수 있게 하는 것 -> 기능의 내부 로직이 바뀌어도 사용하는 측에서는 신경쓸 필요가 없게 해준다. 
Polymorphism:  부모 객체의 기능을 오버라이딩해서 쓸 수 있는것.  ( 단, 해당 기능의 기본 규칙은 지키는 선에서)

**Comment by WooGie911** [comment]
Object-Oriented Programming ( 객체 지향 프로그래밍 ) 의 특징 4가지

1. Abstraction ( 추상화 )
 - 객체 지향 프로그래밍에서 추상화는 데이터 추상화와 제어 추상화로 나뉜다.
 - 데이터 추상화란, 대상을 간단한 개념으로 일반화 하는 과정으로 객체들을 공통 특징으로 묶어서 이름을 붙이는 것이다.
 - 제어 추상화란, 어떤 클래스의 메소드의 작동 방식과 같은 내부 로직은 숨기고 결과만 보이게 하는것이다.
 - 추상화를 통해서 설계를 잘 했다면 생산성 증가, 가독성 증가, 에러 감소, 유지 보수시 시간 단축 등 효과를 얻을 수 있다.

2. Encapsulation ( 캡슐화 )
 - 객체. 즉 클래스의 내부 변수와 메소드를 하나로 패키징 하는 특징.
 - 객체의 응집도와 독립성을 높임으로써 객체의 모듈화를 지향한다.
 - 객체의 모듈화가 잘 이루어져 있을 경우 모듈 단위의 재사용성이 높아져 유지보수에 용이하다.

3. Inheritance ( 상속 )
 - 객체 지향 프로그래밍에선 객체가 다른 객체를 상속받아 상속받은 객체의 요소를 사용 할 수 있다.
 - 상속을 통해 재사용성과 확장성을 보장 받을 수 있다.

4.  Polymorphism ( 다형성 )
 - 다형성이란, 하나의 객체 혹은 메소드가 여러 타입을 참조할수 있음을 의미한다. 
 - 동일한 메소드로 여러 타입의 데이터를 처리하거나, 공통 상속된 객체를 처리함으로써 중복된 코드 소요를 제거하고 개발 편의성을 높여준다.
 - 다형성을 활용함으로써 중복된 코드를 줄이고, 데이터 처리의 범위는 넓힐 수 있다.

**Comment by skyhanull** [comment]
Abstraction : 불필요한 정보는 숨기고 중요한(필요한) 정보만을 표현함으로써 공통의 속성이나 행위(methods)를 하나로 묶어 이름을 붙이는 것이다.
Encapsulation : 객체(object)의 속성과 행위(method)를 하나로 묶고, 구현된 일부를 외부에 감추어 은닉한다.
Inheritance :부모 class의 속성과 행위(methods)을 그대로 상속 받고 행위(methods)의 일부분을 수정해야 할 경우 상속받은 자식 class에서 해당 행위(methods)만 다시 수정(overriding)하여 사용할 수 있도록 한다. 또한 자식 class에서 추가적으로 속성이나 행위(methods)를 정의할 수 있게 한다.
Polymorphism : 여러 형태를 가질 수 있도록 한다. 즉, 객체를 부품화할 수 있도록 한다.

**Comment by BaekYuri** [comment]
Abstraction: 추상화. 어떤 객체의 행동(메소드)이나 속성(변수) 등을 미리 추상화한다. 메소드나 변수명을 보면 어떤 행동을 하는 것인지 알 수 있도록 한다. 
Encapsulation: 캡슐화. 어떤 객체의 행동과 속성을 캡슐화한다.
Inheritance: 상속. 어떤 객체는 부모 객체와의 is a 관계를 가진다. 예를들어 고양이는 **동물**이다. 라고 생각할 때 class Cat extends Animal 과 같은 형식으로 구현할 수 있다. 상속을 받음으로써 부모객체가 가지고 있는 행동 및 속성을 물려받을 수 있다.
Polymorphism: 다형성. 어떠한 함수를 다양한 형태로 만들 수 있다. 자바의 경우에는 오버로딩, 오버라이딩을 하면서 자유롭게 형태를 바꿀 수 있다.

**Comment by javascriptbiggosoo** [comment]
### Abstraction

코드의 사용자가 그것을 단순히 이용하는데에 있어 몰라도 되는 구현의 세부 로직은 숨기고 전체적인 그림만 보여주는 것을 말합니다.


### Encapsulation

위험성을 줄이기 위해 의도하지 않은 방식으로 프로퍼티 혹은 state를 변경하지 못하게끔 일부 프로퍼티와 메소드를 외부에서 접근하지 못하도록 하는 것입니다. 


### Inheritance

재사용성을 위해 클래스의 자식 클래스를 만들어 모든 프로퍼티와 메소드를 동일하게 이용할 수 있도록 넘겨주는 것을 말합니다.

### Polymorphism

개별 자식 클래스의 활용성을 늘이기 위해 상속받은 메소드를 덮어쓸 수 있도록 해주는 것을 말합니다.
